lib/bap_image/bap_image.ml

```diff
@@ -349,10 +349,14 @@ module Derive = struct
     | Some x -> Fact.return x
 
   let arch =
-    Fact.require arch >>= fun s -> match Arch.of_string s with
-    | Some s -> Fact.return s
-    | None -> Fact.failf "unknown/unsupported architecture %s" s ()
-
+    Fact.foreach Ogre.Query.(select (from arch)) ~f:ident >>= fun s ->
+    Fact.Seq.reduce ~f:(fun a1 a2 ->
+        if Arch.equal a1 a2 then Fact.return a1
+        else Fact.failf "arch is ambigous" ())
+        (Seq.filter_map ~f:Arch.of_string s) >>= fun a ->
+    match a with
```

Comment:
At least because I have to feed some function to `Seq.reduce`  and therefore it will not look so briefly (
_a-m-b-i-g-u-o-u-s_ ... it's hard!!